Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -34,309 +34,260 @@
-            return false;
-        }
+            if (shouldTerminateIfNeeded())
+                return true;
 
-        m_client.didReceiveHeaderFromMultipart(m_headers);
-        m_state = State::InContent;
-    }
-    // Fallthrough.
-    case State::InContent: {
-        if (m_buffer.isEmpty())
-            return false;
-
-        size_t boundaryStart;
-        size_t lastPartialMatch;
-
-        if (!checkForBoundary(boundaryStart, lastPartialMatch) && boundaryStart == notFound) {
-            // Did not find the boundary start, all data up to the lastPartialMatch is ok.
-            m_client.didReceiveDataFromMultipart(SharedBuffer::create(m_buffer.data(), lastPartialMatch));
-            m_buffer.remove(0, lastPartialMatch);
             return false;
         }
 
-        // There was a boundary start (or end we'll check that later), push out part of the data.
-        m_client.didReceiveDataFromMultipart(SharedBuffer::create(m_buffer.data(), boundaryStart));
-        m_buffer.remove(0, boundaryStart + m_boundary.length());
-        m_state = State::EndBoundary;
+        m_client.didReceiveHeaderFromMultipart(WTFMove(m_headers));
+        m_state = State::WaitingForHeaderProcessing;
+        return true;
     }
-    // Fallthrough.
-    case State::EndBoundary: {
-        if (m_buffer.size() < 2)
-            return false; // Not enough data to check. Return later when there is more data.
-
-        // We'll decide if this is a closing boundary or an opening one.
-        auto* content = m_buffer.data();
-
-        if (content[0] == '-' && content[1] == '-') {
-            // This is a closing boundary. Close down the handler.
-            m_state = State::End;
-            return false;
-        }
-
-        // This was a simple content separator not a closing one.
-        // Go to before the content processing.
-        m_state = State::InBoundary;
-        break;
+    case State::WaitingForHeaderProcessing: {
+        // Wait until completeHeaderProcessing() is called
+        return false;
     }
-    case State::End:
-        // We are done. Nothing to do anymore.
+    case State::WaitingForTerminate: {
+        // Wait until didCompleteMessage() is called
         return false;
-    default:
-        ASSERT_NOT_REACHED();
+    }
+    case State::Terminating: {
+        m_client.didCompleteFromMultipart();
+        m_state = State::End;
         return false;
     }
+    case State::End:
+        return false;
+    }
+
+    return false;
+}
+
+bool CurlMultipartHandle::shouldTerminateIfNeeded()
+{
+    if (m_didCompleteMessage) {
+        m_state = State::Terminating;
+        return true;
+    }
 
-    return true; // There are still things to process, so go for it.
+    return false;
 }
 
-bool CurlMultipartHandle::checkForBoundary(size_t& boundaryStartPosition, size_t& lastPartialMatchPosition)
+std::optional<bool> CurlMultipartHandle::findBoundary(size_t& matchPos, size_t& endPos, bool hasCloseDelimiter)
```

Comment:
It's difficult to understand. How about defining a struct for the return value?

~~~
struct FindBoundaryResult {
    bool isSyntaxError;
    bool hasBoundary;
    size_t processed;
    size_t dataEnd;
};
~~~

Reply:
Fixed